chainerx_cc/chainerx/routines/normalization.cc

```diff
@@ -117,43 +252,68 @@ Array BatchNorm(
-            x.device().GetBatchNormForwardBackward(result.mean, result.var, eps, decay, result.sorted_axis);
-
     const Array& gamma_reshaped = result.gamma;
     const Array& beta_reshaped = result.beta;
+    const Array& mean_reshaped = result.mean;
+    const Array& var_reshaped = result.var;
+    const Axes& sorted_axis = result.sorted_axis;
+
+    Device& device = x.device();
 
-    Array out = fb->Forward(x.AsGradStopped(), gamma_reshaped.AsGradStopped(), beta_reshaped.AsGradStopped());
-    CHAINERX_ASSERT(out.dtype() == x.dtype());
+    // Compute forward.
+    Array out = EmptyLike(x, device);
+    // Create an empty state container that may be modified during forward. It will then be passed to the backward.
+    nonstd::optional<std::shared_ptr<void>> state{nullptr};
+    {
+        NoBackpropModeScope scope{};
+        device.backend().CallOp<BatchNormForwardOp>(
+                x.AsGradStopped(),
+                gamma_reshaped.AsGradStopped(),
+                beta_reshaped.AsGradStopped(),
+                mean_reshaped,
+                var_reshaped,
+                eps,
+                decay,
+                sorted_axis,
+                out,
+                state);
+    }
     internal::MakeViewForForwardBackwardOutput(out);
 
     BackwardBuilder bb{"batch_norm", {x, gamma_reshaped, beta_reshaped}, {out}};
     if (BackwardBuilder::Target bt = bb.CreateTarget({0, 1, 2})) {
-        bt.Define([fb = std::move(fb), x_tok = bb.RetainInput(0), gamma_tok = bb.RetainInput(1), eps, sorted_axis = result.sorted_axis](
-                          BackwardContext& bctx) {
+        bt.Define([state = std::move(state),
+                   x_tok = bb.RetainInput(0),
+                   gamma_tok = bb.RetainInput(1),
+                   eps,
+                   sorted_axis,
+                   beta_shape = beta_reshaped.shape(),
+                   beta_dtype = beta_reshaped.dtype()](BackwardContext& bctx) mutable {
             const Array& gout = *bctx.output_grad();
+            const Array& x = bctx.GetRetainedInput(x_tok);
+            const Array& gamma_reshaped = bctx.GetRetainedInput(gamma_tok);
 
-            Array gx{};
-            Array ggamma{};
-            Array gbeta{};
-            {
-                std::array<Array, 3> ginputs = fb->Backward(gout.AsGradStopped());
-                internal::MakeViewForForwardBackwardOutput(ginputs);
+            Device& device = x.device();
 
-                gx = std::move(ginputs[0]);
-                ggamma = std::move(ginputs[1]);
-                gbeta = std::move(ginputs[2]);
+            // Compute backward.
+            Array gx = EmptyLike(x, device);
+            Array ggamma = EmptyLike(gamma_reshaped, device);
+            Array gbeta = Empty(beta_shape, beta_dtype, device);
+            {
+                NoBackpropModeScope scope{};
+                device.backend().CallOp<BatchNormBackwardOp>(x, gamma_reshaped, gout, eps, sorted_axis, gx, ggamma, gbeta, state);
```

Comment:
Why passing `gx, ggamma, gbeta` as parameters instead of returning?

Reply:
For the same reason as accepting `out` as an argument and for consistency for `Call`.

Reply:
Changed according to offline discussions:
1. `Op::Call` should return Arrays.
2. `Op::Call` should take output Arrays as optional arguments too.

Reply:
The motivation behind the conclusion was that it's flexible from a caller perspective. The op implementer needs to write more code however. Note that is one may simply compute the output and before retuning it, copy it to the given output parameter if it would be given.